[CORE-14665] security/audit: Fix stuck misconfigured state

IoannisRP opened the pull request:
Fixes: [CORE-14665](https://redpandadata.atlassian.net/browse/CORE-14665)

There are 2 separate fixes in this RP:
- Add post op nagging to the audit log allow list
- Based on the last-error logic controlling the misconfigured state, an illegal_sasl_state error followed by a broker_not_avaible would lock the audit log manager to misconfigured state, at least until a new illegal_sasl_state error would be encountered.
Updated logic to revert the misconfigured on the first successful error_code, regardless of the previous state.

<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [x] v25.3.x
- [x] v25.2.x
- [x] v25.1.x
- [ ] v24.3.x

## Release Notes

* none

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->


[CORE-14665]: https://redpandadata.atlassian.net/browse/CORE-14665?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

IoannisRP committed:
security/audit: Fix stuck misconfigured state

Based on the last-error logic controlling the misconfigured state, an
illegal_sasl_state error followed by a broker_not_avaible would lock the
audit log manager to misconfigured state, at least until a new
illegal_sasl_state error would be encountered.

Updated logic to revert the misconfigured on the first successful
error_code, regardless of the previous state.

IoannisRP commented:
/ci-repeat 1
dt-repeat=50
tests/rptest/tests/audit_log_test.py::AuditLogTestKafkaApi.test_no_auth_enabled

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--adbc5c47-a2e8-4857-9e73-7c36fba02028-->
<!--TEST RESULTS-->
<!--72ffd5e4-4ec7-4873-9529-79cceea85af2-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/76576">build#76576</a></summary>


| test_class           | test_method          | test_arguments                      | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  | test_history                                                                                                                                            |
|:---------------------|:---------------------|:------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:--------------------------------------------------------------------------------------------------------------------------------------------------------|
| AuditLogTestKafkaApi | test_no_auth_enabled | {"audit_transport_mode": "kclient"} | integration | https://buildkite.com/redpanda/redpanda/builds/76576#019a985c-652a-4753-a989-c56d19319226 | FLAKY         | 67/70    | upstream reliability is '98.62595419847328'. current run reliability is '86.36363636363636'. drift is 12.26232 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=AuditLogTestKafkaApi&test_method=test_no_auth_enabled |
| AuditLogTestKafkaApi | test_no_auth_enabled | {"audit_transport_mode": "kclient"} | integration | https://buildkite.com/redpanda/redpanda/builds/76576#019a9888-50e2-4671-ac9e-56099da37140 | FLAKY         | 58/70    | upstream reliability is '98.64457831325302'. current run reliability is '57.14285714285714'. drift is 41.50172 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=AuditLogTestKafkaApi&test_method=test_no_auth_enabled |
| AuditLogTestKafkaApi | test_no_auth_enabled | {"audit_transport_mode": "rpc"}     | integration | https://buildkite.com/redpanda/redpanda/builds/76576#019a9888-50e2-4671-ac9e-56099da37140 | FLAKY         | 68/70    | upstream reliability is '100.0'. current run reliability is '90.47619047619048'. drift is 9.52381 and the allowed drift is set to 50. The test should PASS              | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=AuditLogTestKafkaApi&test_method=test_no_auth_enabled |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/76621">build#76621</a></summary>


| test_class                          | test_method                                  | test_arguments                    | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                   | test_history                                                                                                                                                                               |
|:------------------------------------|:---------------------------------------------|:----------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ConsumerOffsetsConsistencyTest      | test_flipping_leadership                     | null                              | integration | https://buildkite.com/redpanda/redpanda/builds/76621#019a9bee-0971-4120-a66d-72d3de993dc8 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS                | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ConsumerOffsetsConsistencyTest&test_method=test_flipping_leadership                      |
| DataMigrationsApiTest               | test_creating_and_listing_migrations         | null                              | integration | https://buildkite.com/redpanda/redpanda/builds/76621#019a9bf1-9aea-41a5-a66c-3080f53ecbf8 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS                | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DataMigrationsApiTest&test_method=test_creating_and_listing_migrations                   |
| LogCompactionTxRemovalTest          | test_tx_control_batch_removal                | null                              | integration | https://buildkite.com/redpanda/redpanda/builds/76621#019a9bee-0977-4e76-96ea-1ad91f8dda4a | FLAKY         | 6/21     | upstream reliability is '65.807962529274'. current run reliability is '28.57142857142857'. drift is 37.23653 and the allowed drift is set to 50. The test should PASS    | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=LogCompactionTxRemovalTest&test_method=test_tx_control_batch_removal                     |
| LogCompactionTxRemovalUpgradeTest   | test_tx_control_batch_removal_with_upgrade   | {"test_case_name": "All commits"} | integration | https://buildkite.com/redpanda/redpanda/builds/76621#019a9bee-097a-4d2f-a4c4-658db7b3ada4 | FLAKY         | 8/21     | upstream reliability is '77.49287749287748'. current run reliability is '38.095238095238095'. drift is 39.39764 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=LogCompactionTxRemovalUpgradeTest&test_method=test_tx_control_batch_removal_with_upgrade |
| SegmentMsTest                       | test_segment_rolling_with_retention_consumer | null                              | integration | https://buildkite.com/redpanda/redpanda/builds/76621#019a9bf1-9aec-4655-9ba5-16c9fb47ce35 | FLAKY         | 20/21    | upstream reliability is '92.10716871832005'. current run reliability is '95.23809523809523'. drift is -3.13093 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=SegmentMsTest&test_method=test_segment_rolling_with_retention_consumer                   |
| WriteCachingFailureInjectionE2ETest | test_crash_all                               | {"use_transactions": false}       | integration | https://buildkite.com/redpanda/redpanda/builds/76621#019a9bf1-9aec-4655-9ba5-16c9fb47ce35 | FLAKY         | 16/21    | upstream reliability is '91.1911911911912'. current run reliability is '76.19047619047619'. drift is 15.00072 and the allowed drift is set to 50. The test should PASS   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=WriteCachingFailureInjectionE2ETest&test_method=test_crash_all                           |
</details>
<!--/body-->

IoannisRP committed:
dt: add logging error in audit log allow list

When a test with misconfigured auditlog tries to print the cluster
config in the post ops, it will complain about the config request not
being properly audited.

Added nag to the audit log allow list.

IoannisRP commented on src/v/security/audit/client.cc:
```diff
@@ -773,23 +774,29 @@ namespace internal {
     }
-    _last_errc = errc;
+
+    constexpr auto failed_codes = std::to_array(
+      {kafka::error_code::illegal_sasl_state,
+       kafka::error_code::broker_not_available});
```
this is a std::array. There are no allocation involved.

BenPope commented on src/v/security/audit/client.cc:
```diff
@@ -244,10 +244,11 @@ class kafka_client_impl final : public audit_client {
     kafka_sink_impl* sink();
 
-    kafka::error_code _last_errc{kafka::error_code::unknown_server_error};
+    bool _misconfigured{false};
```
You could keep the type safety of `auth_misconfigured_t`.

IoannisRP commented:
changes in [force-push](https://github.com/redpanda-data/redpanda/compare/f9e80f56f267d84755adb9f1aa8c669f932b1e59..24f1ed605fa51cfcdc6b2237913082426f1219f4):
- keep strong typedef for `_misconfigured`

vbotbuildovich commented:
/backport v25.3.x

vbotbuildovich commented:
/backport v25.2.x

vbotbuildovich commented:
/backport v25.1.x

vbotbuildovich commented:
Failed to create a backport PR to v25.1.x branch. I tried:

```
git remote add upstream https://github.com/redpanda-data/redpanda.git
git fetch --all
git checkout -b backport-pr-28616-v25.1.x-237 remotes/upstream/v25.1.x
git cherry-pick -x afbbd8aeda 24f1ed605f
```

[Workflow run logs](https://github.com/redpanda-data/redpanda/actions/runs/19579902675).

vbotbuildovich commented:
Failed to create a backport PR to v25.2.x branch. I tried:

```
git remote add upstream https://github.com/redpanda-data/redpanda.git
git fetch --all
git checkout -b backport-pr-28616-v25.2.x-751 remotes/upstream/v25.2.x
git cherry-pick -x afbbd8aeda 24f1ed605f
```

[Workflow run logs](https://github.com/redpanda-data/redpanda/actions/runs/19579902420).
